auto_compressor/src/state_saving.rs

```diff
@@ -0,0 +1,260 @@
+            );
+        }
+
+        // Check that there is no missing level in the database
+        // e.g. if the previous row retrieved was for level 1 and this
+        // row is for level 3 then since the SQL query orders the results
+        // in ascenting level numbers, there was no level 2 found!
+        if prev_seen != level_num - 1 {
+            bail!("Levels between {} and {} are missing", prev_seen, level_num,);
+        }
+
+        // if the level is not empty, then it must have a head!
+        if current_head.is_none() && current_length != 0 {
+            bail!(
+                "Level {} has no head but current length is {} in room {}",
+                level_num,
+                current_length,
+                room_id,
+            );
+        }
+
+        // If the level has more groups in than the maximum then something is wrong!
+        if current_length > max_size {
+            bail!(
+                "Level {} has length {} but max size {} in room {}",
+                level_num,
+                current_length,
+                max_size,
+                room_id,
+            );
+        }
+
+        // Add this level to the level_info vector
+        level_info.push(Level::restore(max_size, current_length, current_head));
+        // Mark the previous level_number seen as the current one
+        prev_seen = level_num;
+    }
+
+    // If we didn't retrieve anything from the database then there is no saved state
+    // in the database!
+    if level_info.is_empty() {
+        return Ok(None);
+    }
+
+    // Return the compressor state we retrieved
+    Ok(Some((last_compressed, level_info)))
+}
+
+/// Save the level info so it can be loaded by the next run of the compressor
+///
+/// # Arguments
+///
+/// * `client`            - A postgres client used to send the requests to the database
+/// * `room_id`           - The room who's saved compressor state we want to save
+/// * `level_info`        - The state that can be used to restore the compressor later
+/// * `last_compressed`   - The last state_group that was compressed. This is needed
+///                         so that the compressor knows where to start from next
+pub fn write_room_compressor_state(
+    client: &mut Client,
+    room_id: &str,
+    level_info: &[Level],
+    last_compressed: i64,
+) -> Result<()> {
+    // The query we are going to build up
+    let mut sql = "".to_string();
+
+    // Go through every level that the compressor is using
+    for (level_num, level) in level_info.iter().enumerate() {
+        // the 1st level is level 1 not level 0, but enumerate starts at 0
+        // so need to add 1 to get correct number
+        let level_num = level_num + 1;
+
+        // delete the old information for this level from the database
+        sql.push_str(&format!(
+            "DELETE FROM state_compressor_state WHERE room_id = {} AND level_num = {};\n",
+            PGEscape(room_id),
+            level_num,
+        ));
```

Comment:
Can we use the proper client functions here? Rather than manually doing the escaping?

Reply:
I'd presumed that PGEscape was written because there weren't any suitable client functions?

Reply:
That said it doesn't really NEED escaping since room_id isn't user set

Reply:
So, PGEscape was originally used because we wrote the SQL to a file rather than sent it to the DB, so we manually had to do the escaping. For stuff that is hitting the DB directly we can just use `client.execute(".. sql ...", &[params])`, which will handle the escaping of the params correctly itself (in fact, it sends the values separately to postgres so avoids needing to do any escaping entirely)

Reply:
Ohh ok that's much nicer!